### PR TITLE
[11/N] Update all_to_all with CPU/CUDA implementations

### PR DESCRIPTION
**Pull request by H-Huang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #87088 Remove ProcessGroupRoundRobin
* #86409 [13/N] Update gather with CPU/CUDA implementations
* #86408 [12/N] Update scatter with CPU/CUDA implementations
* **#86407 [11/N] Update all_to_all with CPU/CUDA implementations**
* #86368 [10/N] Update barrier with CPU/CUDA implementations
* #86166 [9/N] [Dispatchable Collectives] Update reduce_scatter with CPU / CUDA implementations
* #83916 [7/N] [Dispatchable Collectives] Update reduce with CPU / CUDA implementations

### Changes
- Updates for the all_to_all collective

### Context
https://github.com/pytorch/pytorch/issues/86225

### CODE REVIEW

**Comment by H-Huang** [comment]
@H-Huang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D40181611).

File: test/distributed/test_c10d_common.py [hunk 1]
```diff
@@ -1468,10 +1468,13 @@ def _call_collective_with_varying_tensors(self, backend, collective, *args):
...
-        elif collective == dist.reduce_scatter:
+        elif collective == dist.reduce_scatter or collective == dist.all_to_all:
+            # gloo does not support reduce_scatter or all_to_all
             if backend != "gloo":
-                # gloo does not support reduce_scatter
-                collective(tensor, [tensor], *args)
+                if collective == dist.reduce_scatter:
+                    collective(tensor, [tensor], *args)
+                else:
+                    collective([tensor], [tensor], *args)
```
Comment (kwen2501) [inline test/distributed/test_c10d_common.py:1477]: nit: please see my comment in [10/N] to dissolve this if-else block.

**Review by kwen2501 — APPROVED** [review]
LGTM.

File: test/distributed/test_c10d_common.py (hunk 1 above)
Comment (H-Huang) [reply test/distributed/test_c10d_common.py:1477]: Thanks! Agree with this, will update in future PR to avoid rebase

**Pull request closed without merging**